node/network/statement-distribution/src/lib.rs

```diff
@@ -1856,24 +1856,44 @@ impl StatementDistribution {
 						"New active leaf",
 					);
 
-					let session_index = runtime.get_session_index(ctx.sender(), relay_parent).await?;
-					let info = runtime.get_session_info_by_index(ctx.sender(), relay_parent, session_index).await?;
-					let session_info = &info.session_info;
-
-					active_heads.entry(relay_parent)
-						.or_insert(ActiveHeadData::new(session_info.validators.clone(), session_index, span));
-
-					active_heads.retain(|h, _| {
-						let live = !deactivated.contains(h);
-						if !live {
+					let session_index = match runtime.get_session_index(ctx.sender(), relay_parent).await {
+						Ok(session_index) => session_index,
+						Err(error) => {
+							tracing::trace!(
+								target: LOG_TARGET,
+								hash = ?relay_parent,
+								"Couldn't activate leaf: no session index found: {}",
+								error
+							);
+							continue;
+						}
+					};
+					let info = match runtime.get_session_info_by_index(ctx.sender(), relay_parent, session_index).await {
+						Ok(info) => info,
+						Err(error) => {
 							tracing::trace!(
 								target: LOG_TARGET,
-								hash = ?h,
-								"Deactivating leaf",
+								hash = ?relay_parent,
+								"Couldn't activate leaf: no session info found: {}",
+								error
 							);
+							continue;
 						}
-						live
-					});
+					};
+
+					let session_info = &info.session_info;
+					active_heads.entry(relay_parent)
+						.or_insert(ActiveHeadData::new(session_info.validators.clone(), session_index, span));
+				}
+
+				for deactivated in deactivated {
+					if active_heads.remove(&deactivated).is_some() {
+						tracing::trace!(
+							target: LOG_TARGET,
+							hash = ?deactivated,
+							"Deactivating leaf",
+						);
+					}
```

Comment:
I think we can simplify the fix by moving this section above the activated loop. This is equivalent, because we never activate and deactivate a leaf at the same time.

We can revert the rest of the changes, which makes your question obsolete.

Reply:
> We can revert the rest of the changes, which makes your question obsolete.

Actually, no, let's collect all errors to a vec instead.

> I'm not entirely sure if this is the way we're supposed to handle those errors;

The logic of handling errors is on the caller side depending on whether it's `Fatal` or not: https://github.com/paritytech/polkadot/blob/5dac532349e7629032c52deb67b2ee81c77ce8a1/node/network/statement-distribution/src/lib.rs#L1632-L1638

Reply:
> We can revert the rest of the changes, which makes your question obsolete.

Well, not necessarily, since there's still the problem of the rest of elements of `activated` essentially being ignored if only one's missing. The memory leak here was *definitely* not intended, and it was a direct result of the early return, so it would make sense that the other elements of `activated` being ignored were also a mistake.

Reply:
> Actually, no, let's collect all errors to a vec instead.

Sounds good to me.

Reply:
> Well, not necessarily, since there's still the problem of the rest of elements of activated essentially being ignored if only one's missing.

I just realized that `activated` actually only consists of t most 1 head except for the startup. So the simplest fix would be just moving deactivation handling up.

Reply:
> > Well, not necessarily, since there's still the problem of the rest of elements of activated essentially being ignored if only one's missing.
> 
> I just realized that `activated` actually only consists of t most 1 head except for the startup. So the simplest fix would be just moving deactivation handling up.

So just for my own reference - are you saying that it's fine to ignore rest of the elements here, or that there won't ever be any extra elements?

(If it's the latter then it'd be nice to add a `debug_assert!` here.)

Reply:
Currently we create active leaves only on block import (1 activated head): https://github.com/paritytech/polkadot/blob/5dac532349e7629032c52deb67b2ee81c77ce8a1/node/overseer/src/lib.rs#L727-L758 and on startup only once https://github.com/paritytech/polkadot/blob/c0387bab91c324bb342e2c5de3910ad6d2d1b91b/node/service/src/lib.rs#L678-L680

> (If it's the latter then it'd be nice to add a debug_assert! here.)

We can change the signature to of `activated` instead to be a single entry.

Reply:
> So just for my own reference - are you saying that it's fine to ignore rest of the elements here, or that there won't ever be any extra elements?

It's fine to ignore extra elements on startup for statement-distribution and there won't ever be any extra elements after that.

Reply:
Okay, I've moved the deallocation to the top.

I'm still not exactly comfortable with leaving the loop below as-is; if there's going to be only one element then that definitely should be enforced by the type system, but as you've said that is out of the scope of this PR.

Reply:
Thanks, I'll handle that as part of #2812.